// Copyright 2022 tison <wander4096@gmail.com>.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use thiserror::Error;

pub type ParseResult<T> = Result<T, ParseError>;

#[derive(Debug, Error)]
pub enum ParseError {
    #[error("EndOfStream")]
    EndOfStream,
    #[error("OtherParseError({0})")]
    Other(String),
}

#[derive(Debug, Error)]
pub enum Error {
    #[error(transparent)]
    Io(#[from] std::io::Error),
    #[error("InternalError({0}")]
    Internal(String),
    #[error("ServerError({0})")]
    Server(String),
}

impl From<std::string::FromUtf8Error> for ParseError {
    fn from(e: std::string::FromUtf8Error) -> Self {
        Self::Other(e.to_string())
    }
}

impl From<std::num::ParseFloatError> for ParseError {
    fn from(e: std::num::ParseFloatError) -> Self {
        Self::Other(e.to_string())
    }
}

impl<E: Into<ParseError>> From<E> for Error {
    fn from(e: E) -> Self {
        match e.into() {
            ParseError::EndOfStream => unreachable!("EndOfStream should be handled internally."),
            ParseError::Other(reason) => Self::Internal(reason),
        }
    }
}
